### ISSUE TITLE
[Bug]: Worker stopped processing jobs, and mostly delayed Jobs

### ISSUE THREAD

**Issue by wernermorgenstern** [taskforcesh/bullmq#2466]
### Version

v5.4.2

### Platform

NodeJS

### What happened?

We have a service, where a worker runs, and processes jobs.  After the processing is done, it will create another job, which is delayed (around 64 minutes).
Today, I noticed that the service and worker stopped processing jobs.  There were no error messages in the logs.  When I used BullBoard (I use it as a UI to see jobs), I saw the jobs were still in the delayed state, and like 24 hours overdue.

When I restarted the service, and the worker started, it immediately started processing those delayed jobs.
This is not the first it happened.  Today I though first checked the delayed jobs.

In today's incident, the service has been running for 4 days.

We run in EKS on AWS (NodeJS service, using Typescript).  I use BullMQ Pro. And we are using Groups and each Group has a concurrency set to 1.

### How to reproduce.

I don't have any test code for this

### Relevant log output

```shell
No Logs or error logs were produced
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by manast** [comment]
Do you know if the worker is still connected to Redis when this happens? for instance, does the workers appear list of workers that are online for the queue?

**Comment by wernermorgenstern** [comment]
That I need to check.  That is a good idea.  Is there a method which shows that?  Or something I can look directly in Redis?

**Comment by manast** [comment]
You can use https://api.docs.bullmq.io/classes/v5.Queue.html#getWorkers or Taskforce.sh, I don't know if BullBoard can also show this information.

**Comment by manast** [comment]
Also important to read this if you haven't already: https://docs.bullmq.io/guide/going-to-production#maxretriesperrequest

**Comment by wernermorgenstern** [comment]
> Also important to read this if you haven't already: https://docs.bullmq.io/guide/going-to-production#maxretriesperrequest

I have set that one as null

```
    const redis = new Redis(config.redisBullUrl as string, {
      keepAlive: 1,
      autoResubscribe: true,
      autoResendUnfulfilledCommands: true,
      connectionName: `${config.serviceName}-${config.agentId}-${redisConnection}-bullmq`,
      connectTimeout: helpers.period('10s'),
      enableOfflineQueue: true,
      enableReadyCheck: true,
      enableAutoPipelining: true,
      maxRetriesPerRequest: null,
      retryStrategy: function (times) {
        return Math.max(Math.min(Math.exp(times), 20000), 1000)
      },
      offlineQueue: true
    })
```

**Comment by wernermorgenstern** [comment]
> getWorkers

I don't think Bullboard shows it.  I haven't see it.

For Taskforce.sh, is there a trial version?  For how long?
And what is the cost of it?  I want to see if it might be better than Bullboard.

We use Redis Cloud (we don't self-host our Redis Instances)

**Comment by wernermorgenstern** [comment]
So regarding checking if worker is connected.
I actually have that already in the Code, for me to make API Calls to check that.  So that is good.
So next time we have a similar issue, and I notice the worker is not connected, what would the next steps be in troubleshooting, and resolving this issue?

**Comment by wernermorgenstern** [comment]
I also was reading the one link you gave.  I need to change this `enableOfflineQueue` to false, for the Redis Instance for Queue, and add a job.  But leave it for true for Workers.  So that is one good suggestion

**Comment by manast** [comment]
> For Taskforce.sh, is there a trial version? For how long?
> And what is the cost of it? I want to see if it might be better than Bullboard.

Yes there is trial. You can find the pricing on the webpage: https://taskforce.sh/

**Comment by manast** [comment]
> So next time we have a similar issue, and I notice the worker is not connected, what would the next steps be in troubleshooting, and resolving this issue?

If you have the suggested settings, the workers should automatically reconnect as soon they can, so you should not get this issue anymore.

**Comment by wernermorgenstern** [comment]
With taskforce.sh, I am a bit confused what the difference is between the plans, especially the connections between 1, 5 and 15

**Comment by wernermorgenstern** [comment]
Also, we use redis cloud.  Three instances, one is development, then qa test and then currently one production. So does that mean we need 3 connections?

Also, can we used direct connection or taskforce.sh connector?
what about the pro connector?  Do we need that one?

I will need to make also sure we enable tls on the redis cloud instances

**Comment by manast** [comment]
You need a connection for every host that you want to debug/monitor, so in your example it would be 3 as you suspected. The direct connector can be used if you have access to your Redis host from the internet. If not, you can use the connector. For example the connector is always used for local connections.

**Comment by mistersingh179** [comment]
@manast I am noticing the same issue with a bunch of delayed jobs not getting picked up by the worker.
I do notice that they stay in the delayed job forever till a new job comes in to active state and then the delayed jobs all get get picked up.

![image](https://github.com/taskforcesh/bullmq/assets/48457746/cc66c19f-8c90-4042-8e7b-a14608f19e2b)

Here you can see that job was created 5 mins ago with delay of 1-2 seconds and the job is just sitting there.

![image](https://github.com/taskforcesh/bullmq/assets/48457746/c8d18005-2726-47e8-abb0-f1fee1f7a10e)

Here you can see that the worker is connected.

**Comment by manast** [comment]
What are your worker settings?

**Comment by manast** [comment]
Another thing, do you know why these jobs have such high "started" numbers?

**Comment by manast** [comment]
Also, how are those delayed jobs produced?

**Comment by wernermorgenstern** [comment]
We just had a similar issue this morning.

Here is how we can create the delayed Jobs.

```
const createNextJob = async (
  jobRotateQueue: QueuePro,
  tid: string,
  devCreated: any,
  devConfig: any,
  delay: string
) => {
  const jobData = {
    tid,
    devCreated,
    devConfig
  }

  const jobOptions: JobsProOptions = {
    removeOnComplete: {
      count: 1000
    },
    removeOnFail: {
      count: 10
    },
    attempts: 6,
    delay: 0,
    group: {
      id: tid
    },
    delay: helpers.period(delay) // This converts e.g. '2m' to milliseconds
  }

  try {
    return await jobRotateQueue.add('jobName', jobData, jobOptions)
  } catch (err: any) {
    winstonLogger.error('Unable to add Job', {
        groupId: tid,
        error: { message: err.message, stack: err.stack, code: err.code }
      })
    }
  }
}

```

**Comment by mistersingh179** [comment]
> What are your worker settings?

my worker options are:

```
{
    connection: redisClient,
    concurrency: Number(10),
    limiter: {
      max: 100,
      duration: 1000,
    },
    autorun: false,
    metrics: {
      maxDataPoints: MetricsTime.TWO_WEEKS,
    },
  },
```

with redisClient being

```
const redisClient = new Redis(REDIS_URL, {
  maxRetriesPerRequest: null,
  enableReadyCheck: false,
  retryStrategy: function (times: number) {
    return Math.max(Math.min(Math.exp(times), 20000), 1000);
  },
});

// this prevents node warning which comes from many workers adding listeners to the same redis connection
redisClient.setMaxListeners(100);
```

**Comment by mistersingh179** [comment]
> Also, how are those delayed jobs produced?

When i add a job, in its processor function i try to get a redis lock on a custom key, if it fails to get a lock, i move the job to delayed like this

```
await job.moveToDelayed(Date.now() + randomInt(1000, 2000), token);
throw new DelayedError();
```

**Comment by mistersingh179** [comment]
> Another thing, do you know why these jobs have such high "started" numbers?

they have a high `started` number, because they have been tried many times and every time they were unable to get a lock they were moved to delayed.

**Comment by roggervalf** [comment]
hi @wernermorgenstern, I can see that your are using groups but you are pointing to 5.4.2 bullmq version, not the actual bullmq-pro version. v7.3.0 that was released today is using bullmq 5.4.2. Just for curiosity, are you using bullmq along side with bullmq-pro?

**Comment by wernermorgenstern** [comment]
I am actually using pro.  So do I need to remove the bullmq from the package for?  I use the pro versions of the functions and constructors

**Comment by roggervalf** [comment]
hey, if you are meaning that you have bullmq-pro and bullmq in your packages.json, yes you should only have pro version, as we used fixed versions of bullmq in pro version

**Comment by manast** [comment]
another question, which versión of Redis are you using?

**Comment by wernermorgenstern** [comment]
> hey, if you are meaning that you have bullmq-pro and bullmq in your packages.json, yes you should only have pro version, as we used fixed versions of bullmq in pro version

I will remove and try that

**Comment by wernermorgenstern** [comment]
> another question, which versión of Redis are you using?

Hi, we are using Redis Cloud Enterprise, which is on 7.2

**Comment by roggervalf** [comment]
hi @wernermorgenstern could you pls try upgrading to pro version 7.3.0, that's our last reléase https://github.com/taskforcesh/bullmq/blob/master/docs/gitbook/bullmq-pro/changelog.md#730-2024-03-16 that contains a fix that afects delayed jobs

**Comment by wernermorgenstern** [comment]
> hi @wernermorgenstern could you pls try upgrading to pro version 7.3.0, that's our last reléase https://github.com/taskforcesh/bullmq/blob/master/docs/gitbook/bullmq-pro/changelog.md#730-2024-03-16 that contains a fix that afects delayed jobs

I will do that.  And deploy it to our production environment next week.  What we saw yesterday, that we had one job stuck in active state.   In Taskforce.sh (Trial version) - I love it so far - I saw it had workers, and idle time was between 1s and 10s.  So the worker was still processing other jobs.
But the active job was stuck in active.  I could not even delete it with Taskforce.sh
Only way I could get it unstuck is to restart the service, which created a new worker, and then create a job with the same Group ID, and then the stuck job went out of active state.

**Comment by manast** [comment]
If the job is stuck in active state then it is because the worker is still processing it, usually this is a bug in the processor code, some promise that never resolved for some reason.

**Comment by wernermorgenstern** [comment]
Ok, here is another one.  These are the details:
```
{
  "connection": {},
  "group": {
    "id": "yzDyS1Qp3k",
    "maxSize": 5000
  },
  "backoff": {
    "delay": 5000,
    "type": "exponential"
  },
  "streams": {
    "events": {
      "maxLen": 10000
    }
  },
  "removeOnFail": {
    "count": 10
  },
  "removeOnComplete": {
    "count": 1000
  },
  "delay": 3796000,
  "attempts": 6
}
```
![Screenshot 2024-03-16 at 23 05 46](https://github.com/taskforcesh/bullmq/assets/39772043/12193361-5b96-4e01-bb4a-300ef9f28d5f)
![Screenshot 2024-03-17 at 07 57 51](https://github.com/taskforcesh/bullmq/assets/39772043/10e0111e-4031-471c-ae2d-ec505d14f50c)


The way I got it cleared, is by delete the Job directly in Redis.  I couldn't get it to clear by restarting the service, or trying to deleting it via Taskforce.sh

I understand, Taskforce.sh won't delete active jobs, due to checks.


I will look in the code, maybe there is an unresolved promise.

One thing to note:
1.  It uses Batches
2.  I use a Function, not sandbox (There is a feature request to be able to use sandboxed workers with Batches).

**Comment by manast** [comment]
But other jobs were processing normally, it is just one that got stuck in active while others would continue processing after this one, or?

**Comment by wernermorgenstern** [comment]
Yes that is correct. Other delayed jobs are still processing.  Just that one is stuck in active

**Comment by manast** [comment]
Then I am quite confident the issue is in the processor code, the job is never completing, so I advise to revise your code to make sure that is the case, it usually helps to put some extra logs to see how long the job went through.

**Comment by wernermorgenstern** [comment]
Ok, I will add some extra logs in there.  I noticed also this morning, that the 3 jobs which were in Active State, had a Stalled Counter of 1 (when I checked in Redis).  And there was a lock key in redis as well with the Job Id.

**Comment by manast** [comment]
Not sure what "stalled counter" means in this context. Jobs are marked as stalled and then cleaned by the workers that are working on them by default 30 seconds, this is the normal worker operation.

**Comment by wernermorgenstern** [comment]
So I have a question.  I see that for the Jobs for this specific queue, the returnValue is null.
Even though I do return a hash (msg and duration) as keys.  And when there is an error, I use 
```
job.setAsFailed(err)
throw err
```
(This is in the function, which is per Job.

However, I am using batches, and for the main Handler Function, when it processes the batch, there is no return Value.

How does BullMQ Pro handle the return Value, when the processing is done in a batch?
Here are my worker settings:
```
  rotateWorkerOptions: <WorkerProOptions>{
    concurrency: 4,
    batch: { size: 40 },
    group: {
      concurrency: 1
    },
    useWorkerThreads: false,
    drainDelay: 5,
    maxStalledCount: 10
  },
```

Could it be, that I see stuck in Active Jobs, that I don't have a explicit return out of the main Handler Function?

This is the code for the Worker Handler Function
(mqttClient is the connection to a MQTT Broker)
```
  try {
    const batch = jobBatch.getBatch()
    const listOfPromises = batch.map((job) =>
      processRotateJob(job, mqttClient)
    )
    const responses = await Promise.all(listOfPromises)
  } catch (err: any) {
    winstonLogger.error('Unable to get Batch', {
      jobBatchName: jobBatch.name || undefined,
      jobBatchId: jobBatch?.id || undefined,
      jobGIDId: jobBatch?.gid || undefined,
      jobBatch: {
        attemptsMade: jobBatch.attemptsMade,
        attemptsStarted: jobBatch.attemptsStarted,
        delay: jobBatch.delay,
        failedReason: jobBatch.failedReason,
        finishedOn: jobBatch.finishedOn,
        name: jobBatch.name,
        opts: jobBatch.opts,
        processedBy: jobBatch.processedBy,
        processedOn: jobBatch.processedOn
      },
      error: { message: err.message, stack: err.stack, code: err.code }
    })
    if (mqttClientObject) {
      await mqttClientObject.end(false)
    }
    throw err
  }

  if (mqttClientObject) {
    await mqttClientObject.end(false)
  }
}
```

**Comment by wernermorgenstern** [comment]
Hi, I was wondering, to monitor stuck active jobs, is there a method which returns active jobs, and then the `processedOn` datestamp?  That is what I see.  It is in active state, but there is a `processedOn` timestamp.

So I would like to do an alert for this, and I can create an API Call around, or a monitor for it.

**Comment by manast** [comment]
@wernermorgenstern the best is to put more logs in your jobs, either using BullMQ's job log or just printing out the logs. The idea is to try to see how long has the job managed to proceed and figure out where in your code it got stuck.

**Comment by jesusvalle** [comment]
Hello,

I have been experiencing this issue (delayed jobs completely frozen, and all of them start working again as soon as I push a new one) since I updated bullmq from version 5.1.8 to 5.4.5. I've downgraded to 5.1.8, and it seems that everything is working well again.

**Comment by manast** [comment]
> Hello,
> 
> I have been experiencing this issue (delayed jobs completely frozen, and all of them start working again as soon as I push a new one) since I updated bullmq from version 5.1.8 to 5.4.5. I've downgraded to 5.1.8, and it seems that everything is working well again.

What options do you have on your workers?

**Comment by jesusvalle** [comment]
> > Hello,
> > I have been experiencing this issue (delayed jobs completely frozen, and all of them start working again as soon as I push a new one) since I updated bullmq from version 5.1.8 to 5.4.5. I've downgraded to 5.1.8, and it seems that everything is working well again.
> 
> What options do you have on your workers?

Here you go! (we are using bullmq with nestjs, btw)

`@Processor(queueName, {
    concurrency: 1,
    autorun: true,
    limiter: {
        max: 60,
        duration: 1000
    }
})`

**Comment by roggervalf** [comment]
Hi @jesusvalle could you pls try version 5.1.10. I would like to discriminate the culprit versión.

**Comment by jesusvalle** [comment]
> Hi @jesusvalle could you pls try version 5.1.10. I would like to discriminate the culprit versión.

Hello Rogger,

We have tested it and it seems that with 5.1.10 it's not ocurring.

With version 5.5.0 the bug still happening.

**Comment by roggervalf** [comment]
hi @jesusvalle in that case it should be safe to upgrade to v5.4.0 as there is not change for markers or blockTimeout values, could you pls try it and let us know

**Comment by manast** [comment]
@jesusvalle yes, it would be very useful to know from which version this is happening, between 5.1 and 5.4 I could not find anything that could affect delayed jobs, but we are still investigating...

**Comment by jesusvalle** [comment]
> hi @jesusvalle in that case it should be safe to upgrade to v5.4.0 as there is not change for markers or blockTimeout values, could you pls try it and let us know

> @jesusvalle yes, it would be very useful to know from which version this is happening, between 5.1 and 5.4 I could not find anything that could affect delayed jobs, but we are still investigating...

Perfect, I will try this version and share the feedback. Thank you guys.

**Comment by manast** [comment]
Next time anybody gets affected by this issue it would be streaming useful if you could issue a redis-cli monitor command to see what is going on in Redis, I am specifically interested in knowing if there are calls to BZPOPMIN at some regular intervals.

**Comment by roggervalf** [comment]
hey guys. we identified that this issue is happening with redis versions lower than 7.0.8 and only happening when minimum block timeout is passed into bzpopmin, that value is 0.001

**Comment by manast** [comment]
Yes, and for versions lower than 7.0.8, we cannot use a timeout lower than 0.002 (2ms)

**Comment by ChrisLahaye** [comment]
Hi @manast! We upgraded from 4.14.0 to 5.7.1 and are now experiencing this exact same bug. Jobs stay stuck in delayed until we restart the worker. Our redis version is 7.1.0.

**Comment by manast** [comment]
> Hi @manast! We upgraded from 4.14.0 to 5.7.1 and are now experiencing this exact same bug. Jobs stay stuck in delayed until we restart the worker. Our redis version is 7.1.0.

Do you have any additional information that can help us investigate the issue? Like I write here for example: https://github.com/taskforcesh/bullmq/issues/2466#issuecomment-2039660908

**Comment by ChrisLahaye** [comment]
> > Hi @manast! We upgraded from 4.14.0 to 5.7.1 and are now experiencing this exact same bug. Jobs stay stuck in delayed until we restart the worker. Our redis version is 7.1.0.
> 
> Do you have any additional information that can help us investigate the issue? Like I write here for example: [#2466 (comment)](https://github.com/taskforcesh/bullmq/issues/2466#issuecomment-2039660908)

Unfortunately I don't have more information. According to bull-board the job is scheduled for a time that is in the past, but it won't run until restarting the worker process. I will roll out the latest version and look for BZPOPMIN commands once the issue starts appearing.

**Comment by ChrisLahaye** [comment]
@manast I have some logs grepped on the queue name `ingress:avalanche`.

```
1713444562.989196 [2 10.0.244.85:50236] "evalsha" "06d4f5e167f90874bcdd2976c56ef657191cbb11" "6" "bull:ingress:avalanche:marker" "bull:ingress:avalanche:meta" "bull:ingress:avalanche:id" "bull:ingress:avalanche:delayed" "bull:ingress:avalanche:completed" "bull:ingress:avalanche:events" "\x99\xb7bull:ingress:avalanche:\xa0\xa1*\xcbBx\xef\x14\x12\x82\xc0\x00\xc0\xc0\xc0\xc0\xc0" "{\"persist\":false,\"meta\":{\"block\":{\"number\":44358230,\"timestamp\":1713444557},\"hasIndexingErrors\":false},\"tokens_createdAt_gt\":\"1694448827\"}" "\xde\x00\x06\xa8attempts\xcb\x7f\xef\xff\xff\xff\xff\xff\xff\xa5delay\xcd\x03\xe8\xa7backoff\xde\x00\x01\xa4type\xa6custom\xb0removeOnComplete\xc3\xacremoveOnFail\xc3\xa5jobId\xc0"
1713444562.989289 [2 lua] "INCR" "bull:ingress:avalanche:id"
1713444562.989307 [2 lua] "HGET" "bull:ingress:avalanche:meta" "opts.maxLenEvents"
1713444562.989350 [2 lua] "HMSET" "bull:ingress:avalanche:149094" "name" "*" "data" "{\"persist\":false,\"meta\":{\"block\":{\"number\":44358230,\"timestamp\":1713444557},\"hasIndexingErrors\":false},\"tokens_createdAt_gt\":\"1694448827\"}" "opts" "{\"backoff\":{\"type\":\"custom\"},\"removeOnFail\":true,\"removeOnComplete\":true,\"delay\":1000,\"attempts\":1.7976931348623e+308}" "timestamp" "1713444562988" "delay" "1000" "priority" "0"
1713444562.989394 [2 lua] "XADD" "bull:ingress:avalanche:events" "*" "event" "added" "jobId" "149094" "name" "*"
1713444562.989413 [2 lua] "ZADD" "bull:ingress:avalanche:delayed" "7018268934096486" "149094"
1713444562.989427 [2 lua] "XADD" "bull:ingress:avalanche:events" "MAXLEN" "~" "10000" "*" "event" "delayed" "jobId" "149094" "delay" "1713444563988"
1713444562.989439 [2 lua] "HEXISTS" "bull:ingress:avalanche:meta" "paused"
1713444562.989451 [2 lua] "ZRANGE" "bull:ingress:avalanche:delayed" "0" "0" "WITHSCORES"
1713444562.989466 [2 lua] "ZADD" "bull:ingress:avalanche:marker" "1713444563988.3999" "1"
1713444562.990414 [2 10.0.244.85:50452] "evalsha" "ac6eda8493b217849bcc9a04125ec35328e13012" "14" "bull:ingress:avalanche:wait" "bull:ingress:avalanche:active" "bull:ingress:avalanche:prioritized" "bull:ingress:avalanche:events" "bull:ingress:avalanche:stalled" "bull:ingress:avalanche:limiter" "bull:ingress:avalanche:delayed" "bull:ingress:avalanche:paused" "bull:ingress:avalanche:meta" "bull:ingress:avalanche:pc" "bull:ingress:avalanche:completed" "bull:ingress:avalanche:149093" "bull:ingress:avalanche:metrics:completed" "bull:ingress:avalanche:marker" "149093" "1713444562989" "returnvalue" "null" "completed" "1" "bull:ingress:avalanche:" "\xde\x00\t\xa5token\xd9+7eaa5009-4183-408b-ac79-7326e73d3c72:210401\xa8keepJobs\xde\x00\x01\xa5count\x00\xa7limiter\xc0\xaclockDuration\xcdu0\xa8attempts\xcb\x7f\xef\xff\xff\xff\xff\xff\xb0\xaemaxMetricsSize\xa0\xa4fpof\xc2\xa4idof\xc2\xa4rdof\xc2"
1713444562.990503 [2 lua] "EXISTS" "bull:ingress:avalanche:149093"
1713444562.990520 [2 lua] "GET" "bull:ingress:avalanche:149093:lock"
1713444562.990531 [2 lua] "DEL" "bull:ingress:avalanche:149093:lock"
1713444562.990538 [2 lua] "SREM" "bull:ingress:avalanche:stalled" "149093"
1713444562.990544 [2 lua] "SCARD" "bull:ingress:avalanche:149093:dependencies"
1713444562.990552 [2 lua] "HMGET" "bull:ingress:avalanche:149093" "parentKey" "parent"
1713444562.990566 [2 lua] "LREM" "bull:ingress:avalanche:active" "-1" "149093"
1713444562.990575 [2 lua] "HGET" "bull:ingress:avalanche:meta" "opts.maxLenEvents"
1713444562.990583 [2 lua] "XTRIM" "bull:ingress:avalanche:events" "MAXLEN" "~" "10000"
1713444562.990593 [2 lua] "HINCRBY" "bull:ingress:avalanche:149093" "atm" "1"
1713444562.990607 [2 lua] "DEL" "bull:ingress:avalanche:149093" "bull:ingress:avalanche:149093:logs" "bull:ingress:avalanche:149093:dependencies" "bull:ingress:avalanche:149093:processed" "bull:ingress:avalanche:149093:failed"
1713444562.990629 [2 lua] "XADD" "bull:ingress:avalanche:events" "*" "event" "completed" "jobId" "149093" "returnvalue" "null"
1713444562.990640 [2 lua] "HEXISTS" "bull:ingress:avalanche:meta" "paused"
1713444562.990655 [2 lua] "ZRANGEBYSCORE" "bull:ingress:avalanche:delayed" "0" "7018268930007040" "LIMIT" "0" "1000"
1713444562.990666 [2 lua] "RPOPLPUSH" "bull:ingress:avalanche:wait" "bull:ingress:avalanche:active"
1713444562.990676 [2 lua] "ZPOPMIN" "bull:ingress:avalanche:prioritized"
1713444562.990681 [2 lua] "DEL" "bull:ingress:avalanche:pc"
1713444562.990689 [2 lua] "ZRANGE" "bull:ingress:avalanche:delayed" "0" "0" "WITHSCORES"
1713444562.991331 [2 10.0.244.85:50454] "bzpopmin" "bull:ingress:avalanche:marker" "0.997"
1713444562.992672 [2 10.0.244.85:50452] "evalsha" "135a97538ee393c145e6d418b575b3aa1799c691" "11" "bull:ingress:avalanche:wait" "bull:ingress:avalanche:active" "bull:ingress:avalanche:prioritized" "bull:ingress:avalanche:events" "bull:ingress:avalanche:stalled" "bull:ingress:avalanche:limiter" "bull:ingress:avalanche:delayed" "bull:ingress:avalanche:paused" "bull:ingress:avalanche:meta" "bull:ingress:avalanche:pc" "bull:ingress:avalanche:marker" "bull:ingress:avalanche:" "1713444562992" "\xde\x00\x04\xa5token\xd9+7eaa5009-4183-408b-ac79-7326e73d3c72:210402\xaclockDuration\xcdu0\xa7limiter\xc0\xa4name\xc0"
1713444562.992751 [2 lua] "HEXISTS" "bull:ingress:avalanche:meta" "paused"
1713444562.992771 [2 lua] "ZRANGEBYSCORE" "bull:ingress:avalanche:delayed" "0" "7018268930019328" "LIMIT" "0" "1000"
1713444562.992783 [2 lua] "RPOPLPUSH" "bull:ingress:avalanche:wait" "bull:ingress:avalanche:active"
1713444562.992792 [2 lua] "ZPOPMIN" "bull:ingress:avalanche:prioritized"
1713444562.992799 [2 lua] "DEL" "bull:ingress:avalanche:pc"
1713444562.992809 [2 lua] "ZRANGE" "bull:ingress:avalanche:delayed" "0" "0" "WITHSCORES"
1713444562.993403 [2 10.0.244.85:50454] "bzpopmin" "bull:ingress:avalanche:marker" "0.995"
1713444562.994151 [2 10.0.244.85:50454] "bzpopmin" "bull:ingress:avalanche:marker" "0.995"
1713444564.003089 [2 10.0.244.85:50452] "evalsha" "135a97538ee393c145e6d418b575b3aa1799c691" "11" "bull:ingress:avalanche:wait" "bull:ingress:avalanche:active" "bull:ingress:avalanche:prioritized" "bull:ingress:avalanche:events" "bull:ingress:avalanche:stalled" "bull:ingress:avalanche:limiter" "bull:ingress:avalanche:delayed" "bull:ingress:avalanche:paused" "bull:ingress:avalanche:meta" "bull:ingress:avalanche:pc" "bull:ingress:avalanche:marker" "bull:ingress:avalanche:" "1713444564002" "\xde\x00\x04\xa5token\xd9+7eaa5009-4183-408b-ac79-7326e73d3c72:210404\xaclockDuration\xcdu0\xa7limiter\xc0\xa4name\xc0"
1713444564.003255 [2 lua] "HEXISTS" "bull:ingress:avalanche:meta" "paused"
1713444564.003298 [2 lua] "ZRANGEBYSCORE" "bull:ingress:avalanche:delayed" "0" "7018268934156288" "LIMIT" "0" "1000"
1713444564.003317 [2 lua] "ZREM" "bull:ingress:avalanche:delayed" "149094"
1713444564.003330 [2 lua] "HGET" "bull:ingress:avalanche:149094" "priority"
1713444564.003344 [2 lua] "LPUSH" "bull:ingress:avalanche:wait" "149094"
1713444564.003356 [2 lua] "ZADD" "bull:ingress:avalanche:marker" "0" "0"
1713444564.003373 [2 lua] "XADD" "bull:ingress:avalanche:events" "*" "event" "waiting" "jobId" "149094" "prev" "delayed"
1713444564.003385 [2 lua] "HSET" "bull:ingress:avalanche:149094" "delay" "0"
1713444564.003399 [2 lua] "RPOPLPUSH" "bull:ingress:avalanche:wait" "bull:ingress:avalanche:active"
1713444564.003417 [2 lua] "SET" "bull:ingress:avalanche:149094:lock" "7eaa5009-4183-408b-ac79-7326e73d3c72:210404" "PX" "30000"
1713444564.003431 [2 lua] "XADD" "bull:ingress:avalanche:events" "*" "event" "active" "jobId" "149094" "prev" "waiting"
1713444564.003468 [2 lua] "HSET" "bull:ingress:avalanche:149094" "processedOn" "1713444564002"
1713444564.003478 [2 lua] "HINCRBY" "bull:ingress:avalanche:149094" "ats" "1"
1713444564.003489 [2 lua] "HGETALL" "bull:ingress:avalanche:149094"
1713444564.157597 [2 10.0.244.85:50236] "evalsha" "06d4f5e167f90874bcdd2976c56ef657191cbb11" "6" "bull:ingress:avalanche:marker" "bull:ingress:avalanche:meta" "bull:ingress:avalanche:id" "bull:ingress:avalanche:delayed" "bull:ingress:avalanche:completed" "bull:ingress:avalanche:events" "\x99\xb7bull:ingress:avalanche:\xa0\xa1*\xcbBx\xef\x14\x12\xcb\xd0\x00\xc0\xc0\xc0\xc0\xc0" "{\"persist\":false,\"meta\":{\"block\":{\"number\":44358230,\"timestamp\":1713444557},\"hasIndexingErrors\":false},\"tokens_createdAt_gt\":\"1694448827\"}" "\xde\x00\x06\xa8attempts\xcb\x7f\xef\xff\xff\xff\xff\xff\xff\xa5delay\xcd\x03\xe8\xa7backoff\xde\x00\x01\xa4type\xa6custom\xb0removeOnComplete\xc3\xacremoveOnFail\xc3\xa5jobId\xc0"
1713444564.157687 [2 lua] "INCR" "bull:ingress:avalanche:id"
1713444564.157699 [2 lua] "HGET" "bull:ingress:avalanche:meta" "opts.maxLenEvents"
1713444564.157741 [2 lua] "HMSET" "bull:ingress:avalanche:149095" "name" "*" "data" "{\"persist\":false,\"meta\":{\"block\":{\"number\":44358230,\"timestamp\":1713444557},\"hasIndexingErrors\":false},\"tokens_createdAt_gt\":\"1694448827\"}" "opts" "{\"backoff\":{\"type\":\"custom\"},\"removeOnFail\":true,\"removeOnComplete\":true,\"delay\":1000,\"attempts\":1.7976931348623e+308}" "timestamp" "1713444564157" "delay" "1000" "priority" "0"
1713444564.157785 [2 lua] "XADD" "bull:ingress:avalanche:events" "*" "event" "added" "jobId" "149095" "name" "*"
1713444564.157803 [2 lua] "ZADD" "bull:ingress:avalanche:delayed" "7018268938884711" "149095"
1713444564.157820 [2 lua] "XADD" "bull:ingress:avalanche:events" "MAXLEN" "~" "10000" "*" "event" "delayed" "jobId" "149095" "delay" "1713444565157"
1713444564.157833 [2 lua] "HEXISTS" "bull:ingress:avalanche:meta" "paused"
1713444564.157844 [2 lua] "ZRANGE" "bull:ingress:avalanche:delayed" "0" "0" "WITHSCORES"
1713444564.157859 [2 lua] "ZADD" "bull:ingress:avalanche:marker" "1713444565157.4001" "1"
1713444564.158798 [2 10.0.244.85:50452] "evalsha" "ac6eda8493b217849bcc9a04125ec35328e13012" "14" "bull:ingress:avalanche:wait" "bull:ingress:avalanche:active" "bull:ingress:avalanche:prioritized" "bull:ingress:avalanche:events" "bull:ingress:avalanche:stalled" "bull:ingress:avalanche:limiter" "bull:ingress:avalanche:delayed" "bull:ingress:avalanche:paused" "bull:ingress:avalanche:meta" "bull:ingress:avalanche:pc" "bull:ingress:avalanche:completed" "bull:ingress:avalanche:149094" "bull:ingress:avalanche:metrics:completed" "bull:ingress:avalanche:marker" "149094" "1713444564158" "returnvalue" "null" "completed" "1" "bull:ingress:avalanche:" "\xde\x00\t\xa5token\xd9+7eaa5009-4183-408b-ac79-7326e73d3c72:210404\xa8keepJobs\xde\x00\x01\xa5count\x00\xa7limiter\xc0\xaclockDuration\xcdu0\xa8attempts\xcb\x7f\xef\xff\xff\xff\xff\xff\xb0\xaemaxMetricsSize\xa0\xa4fpof\xc2\xa4idof\xc2\xa4rdof\xc2"
1713444564.158889 [2 lua] "EXISTS" "bull:ingress:avalanche:149094"
1713444564.158906 [2 lua] "GET" "bull:ingress:avalanche:149094:lock"
1713444564.158915 [2 lua] "DEL" "bull:ingress:avalanche:149094:lock"
1713444564.158922 [2 lua] "SREM" "bull:ingress:avalanche:stalled" "149094"
1713444564.158928 [2 lua] "SCARD" "bull:ingress:avalanche:149094:dependencies"
1713444564.158936 [2 lua] "HMGET" "bull:ingress:avalanche:149094" "parentKey" "parent"
1713444564.158950 [2 lua] "LREM" "bull:ingress:avalanche:active" "-1" "149094"
1713444564.158958 [2 lua] "HGET" "bull:ingress:avalanche:meta" "opts.maxLenEvents"
1713444564.158967 [2 lua] "XTRIM" "bull:ingress:avalanche:events" "MAXLEN" "~" "10000"
1713444564.158977 [2 lua] "HINCRBY" "bull:ingress:avalanche:149094" "atm" "1"
1713444564.158996 [2 lua] "DEL" "bull:ingress:avalanche:149094" "bull:ingress:avalanche:149094:logs" "bull:ingress:avalanche:149094:dependencies" "bull:ingress:avalanche:149094:processed" "bull:ingress:avalanche:149094:failed"
1713444564.159019 [2 lua] "XADD" "bull:ingress:avalanche:events" "*" "event" "completed" "jobId" "149094" "returnvalue" "null"
1713444564.159030 [2 lua] "HEXISTS" "bull:ingress:avalanche:meta" "paused"
1713444564.159045 [2 lua] "ZRANGEBYSCORE" "bull:ingress:avalanche:delayed" "0" "7018268934795264" "LIMIT" "0" "1000"
1713444564.159056 [2 lua] "RPOPLPUSH" "bull:ingress:avalanche:wait" "bull:ingress:avalanche:active"
1713444564.159065 [2 lua] "ZPOPMIN" "bull:ingress:avalanche:prioritized"
1713444564.159070 [2 lua] "DEL" "bull:ingress:avalanche:pc"
1713444564.159078 [2 lua] "ZRANGE" "bull:ingress:avalanche:delayed" "0" "0" "WITHSCORES"
1713444564.159723 [2 10.0.244.85:50454] "bzpopmin" "bull:ingress:avalanche:marker" "0.998"
1713444564.160622 [2 10.0.244.85:50452] "evalsha" "135a97538ee393c145e6d418b575b3aa1799c691" "11" "bull:ingress:avalanche:wait" "bull:ingress:avalanche:active" "bull:ingress:avalanche:prioritized" "bull:ingress:avalanche:events" "bull:ingress:avalanche:stalled" "bull:ingress:avalanche:limiter" "bull:ingress:avalanche:delayed" "bull:ingress:avalanche:paused" "bull:ingress:avalanche:meta" "bull:ingress:avalanche:pc" "bull:ingress:avalanche:marker" "bull:ingress:avalanche:" "1713444564160" "\xde\x00\x04\xa5token\xd9+7eaa5009-4183-408b-ac79-7326e73d3c72:210405\xaclockDuration\xcdu0\xa7limiter\xc0\xa4name\xc0"
1713444564.160688 [2 lua] "HEXISTS" "bull:ingress:avalanche:meta" "paused"
1713444564.160704 [2 lua] "ZRANGEBYSCORE" "bull:ingress:avalanche:delayed" "0" "7018268934803456" "LIMIT" "0" "1000"
1713444564.160715 [2 lua] "RPOPLPUSH" "bull:ingress:avalanche:wait" "bull:ingress:avalanche:active"
1713444564.160723 [2 lua] "ZPOPMIN" "bull:ingress:avalanche:prioritized"
1713444564.160730 [2 lua] "DEL" "bull:ingress:avalanche:pc"
1713444564.160738 [2 lua] "ZRANGE" "bull:ingress:avalanche:delayed" "0" "0" "WITHSCORES"
1713444564.161293 [2 10.0.244.85:50454] "bzpopmin" "bull:ingress:avalanche:marker" "0.996"
1713444564.162011 [2 10.0.244.85:50454] "bzpopmin" "bull:ingress:avalanche:marker" "0.996"
```

Does this help?

**Comment by manast** [comment]
What happens after the last call to bzpopmin ?

**Comment by manast** [comment]
What I mean, does it just get stuck there or more commands are coming afterwards?

**Comment by manast** [comment]
@ChrisLahaye thanks for the info!

**Comment by manast** [comment]
In the logs you pasted above I can see jobs have are been completed actually. Are this logs produced when you say that no jobs are being processed?

**Comment by ChrisLahaye** [comment]
@manast Oops, I accidentally shared the logs for a working queue. The following logs is for a queue which isn't processing jobs.

```
1713519805.515271 [2 10.0.244.85:50398] "evalsha" "4d3d690062d87ee95eb0ee95beddf6b678c9000c" "9" "bull:ingress:optimism:stalled" "bull:ingress:optimism:wait" "bull:ingress:optimism:active" "bull:ingress:optimism:failed" "bull:ingress:optimism:stalled-check" "bull:ingress:optimism:meta" "bull:ingress:optimism:paused" "bull:ingress:optimism:marker" "bull:ingress:optimism:events" "1" "bull:ingress:optimism:" "1713519805514" "30000"
1713519805.515366 [2 lua] "EXISTS" "bull:ingress:optimism:stalled-check"
1713519805.515406 [2 lua] "SET" "bull:ingress:optimism:stalled-check" "1713519805514" "PX" "30000"
1713519805.515422 [2 lua] "HGET" "bull:ingress:optimism:meta" "opts.maxLenEvents"
1713519805.515435 [2 lua] "XTRIM" "bull:ingress:optimism:events" "MAXLEN" "~" "10000"
1713519805.515443 [2 lua] "SMEMBERS" "bull:ingress:optimism:stalled"
1713519805.515456 [2 lua] "LRANGE" "bull:ingress:optimism:active" "0" "-1"

1713519835.516088 [2 10.0.244.85:50398] "evalsha" "4d3d690062d87ee95eb0ee95beddf6b678c9000c" "9" "bull:ingress:optimism:stalled" "bull:ingress:optimism:wait" "bull:ingress:optimism:active" "bull:ingress:optimism:failed" "bull:ingress:optimism:stalled-check" "bull:ingress:optimism:meta" "bull:ingress:optimism:paused" "bull:ingress:optimism:marker" "bull:ingress:optimism:events" "1" "bull:ingress:optimism:" "1713519835515" "30000"
1713519835.516168 [2 lua] "EXISTS" "bull:ingress:optimism:stalled-check"
1713519835.516183 [2 lua] "SET" "bull:ingress:optimism:stalled-check" "1713519835515" "PX" "30000"
1713519835.516197 [2 lua] "HGET" "bull:ingress:optimism:meta" "opts.maxLenEvents"
1713519835.516208 [2 lua] "XTRIM" "bull:ingress:optimism:events" "MAXLEN" "~" "10000"
1713519835.516217 [2 lua] "SMEMBERS" "bull:ingress:optimism:stalled"
1713519835.516228 [2 lua] "LRANGE" "bull:ingress:optimism:active" "0" "-1"
```

This just keeps repeating. I verified the logs by comparing it with another queue which is also stuck and that one shows the same behavior.

**Comment by manast** [comment]
@ChrisLahaye could you also share the result of calling CLIENT LIST in a redis queue that is in this state?

**Comment by ChrisLahaye** [comment]
> @ChrisLahaye could you also share the result of calling CLIENT LIST in a redis queue that is in this state?

Do I need to use redis-cli and then execute the CLIENT LIST redis command? We have quite a lot of different queues and services using this redis instance. Could you provide some more detailed instructions to get the relevant data?

**Comment by manast** [comment]
@ChrisLahaye basically what we want is to see which was the latest command issued in the worker's client connection. Unfortunately the queue name is encoded in base64 in the client name so it is not easy to filter it out. But if you send us the complete output we can figure it out.

**Comment by manast** [comment]
So yes, redis-cli and then CLIENT LIST

**Comment by ChrisLahaye** [comment]
@manast 
```
id=467985 addr=10.0.244.85:50442 laddr=10.0.178.10:6379 fd=88 name=bull:aW5ncmVzczpvcHRpbWlzbQ== age=185012 idle=133054 flags=b db=2 sub=0 psub=0 ssub=0 multi=-1 qbuf=0 qbuf-free=0 argv-mem=41 multi-mem=0 rbs=1024 rbp=0 obl=0 oll=0 omem=0 tot-mem=2609 events=r cmd=bzpopmin user=default redir=-1 resp=2
```

**Comment by manast** [comment]
And as soon as you add a new job to this queue, the worker wakes up right?

**Comment by manast** [comment]
Our suspicion here is that it seems that bzpopmin is issuing a timeout value that Redis considers as timeout indefinitely... we are a bit puzzled by this as we are making sure to always use a value larger than 0.002

**Comment by manast** [comment]
(0.001 in older versions of Redis was considered also as indefinitely timeout as we suspect it was a bug that has been resolved but to be sure we use 0.002 (2ms) minimum timeout)

**Comment by ChrisLahaye** [comment]
> And as soon as you add a new job to this queue, the worker wakes up right?

We haven't tested that. We create the initial job once and then this job keeps rescheduling itself but with different data. We use the following pattern,

```ts
const jobOptions: JobsOptions = {
  attempts: Number.MAX_VALUE,
  backoff: { type: 'custom' },
  delay: POLL_DELAY,
  removeOnComplete: true,
  removeOnFail: true,
};

async function processor({ data }: Job<JobData>) {
...
   await queue.add('*', nextData, jobOptions);
...
}
```

At some point it gets stuck and bull board shows that there is a scheduled job with a timestamp in the past. After restarting the process this job finally runs.

**Comment by manast** [comment]
@ChrisLahaye it would be useful to know if adding a new job to the queue wakes up the worker, this is the last piece in the puzzle that we need to understand. Also, it is very important to verify that the worker that gets stuck actually is running BullMQ's latest version.

**Comment by ChrisLahaye** [comment]
@manast after adding a new job to the queue the worker that is stuck awakes and runs the new job as well as the previously stuck jobs. 

https://gist.github.com/ChrisLahaye/3f38b27d8ab68b2d3ffc2203a6f6d738

**Comment by ChrisLahaye** [comment]
Just to confirm, we are running bullmq version 5.7.1

**Comment by roggervalf** [comment]
hi @ChrisLahaye could you pls use v5.7.3 just in case?

**Comment by ChrisLahaye** [comment]
We just updated to v5.7.3 and will reach out once a worker gets stuck.

**Comment by oanguenot** [comment]
@manast, I confirm on my own that adding a new job wakes up the worker. 
I have upgraded to latest 5.7.3 same as @ChrisLahaye, I will check as soon as I have the issue

**Comment by ChrisLahaye** [comment]
The issue is still happening on v5.7.3

**Comment by roggervalf** [comment]
> @manast after adding a new job to the queue the worker that is stuck awakes and runs the new job as well as the previously stuck jobs.
> 
> https://gist.github.com/ChrisLahaye/3f38b27d8ab68b2d3ffc2203a6f6d738

hi @ChrisLahaye I took a look in your logs but what will help us more is the actual value that is passed into bzpopmin that is causing to be blocked

**Comment by ChrisLahaye** [comment]
> > @manast after adding a new job to the queue the worker that is stuck awakes and runs the new job as well as the previously stuck jobs.
> > 
> > https://gist.github.com/ChrisLahaye/3f38b27d8ab68b2d3ffc2203a6f6d738
> 
> hi @ChrisLahaye I took a look in your logs but what will help us more is the actual value that is passed into bzpopmin that is causing to be blocked

How do I find that value? Currently doing a grep with the queue name over the monitor output. Is the value filtered out or not in this output?

**Comment by roggervalf** [comment]
hey @ChrisLahaye so until line 18 in https://gist.github.com/ChrisLahaye/3f38b27d8ab68b2d3ffc2203a6f6d738 is showing when stalled checks are happening, this operation by default it's executed every 30 seconds, I suppose here is where your worker is stopped, what we would like to know if what is happening before. A call to bzpopmin should happen before getting stopped (only calls to stalled-checks are happening)

**Comment by ChrisLahaye** [comment]
> hey @ChrisLahaye so until line 18 in https://gist.github.com/ChrisLahaye/3f38b27d8ab68b2d3ffc2203a6f6d738 is showing when stalled checks are happening, this operation by default it's executed every 30 seconds, I suppose here is where your worker is stopped, what we would like to know if what is happening before. A call to bzpopmin should happen before getting stopped (only calls to stalled-checks are happening)

What do you recommend me to do? Start logging on a working queue and catching when it breaks?

**Comment by manast** [comment]
@ChrisLahaye in principle, using the latest version you should not get this issue. But if you get it again, then it would be pretty useful to know the arguments on the last call to BZPOPMIN, this is obviously difficult to get because it means you need to have monitor enabled until it happens, which costs quite a lot of CPU if you are processing a lot of jobs, so it may not be feasible in practice. In any case, if it happens again with the latest code, we would need to assume that somehow BZPOPMIN ends up being called with a value that Redis interprets as infinite timeout, and we will need to find a workaround for this. Please keep us posted.

**Comment by ChrisLahaye** [comment]
@manast is v5.7.4 supposed to fix the issue?

**Comment by manast** [comment]
5.7.1 for Node and 5.7.2 for Python.

**Comment by ChrisLahaye** [comment]
Unfortunately we are still experiencing the issue with the latest version. Could you please let me know which version introduced this issue so we can downgrade to that specific version.

**Comment by matthewgonzalez** [comment]
Hi, we've experience this 3 times in the past couple of weeks on a critical queue that repeats every 30 seconds.

I've bumped Bull to `5.7.1` with no luck. We are on Redis 6.2.6.

**Comment by roggervalf** [comment]
hi guys, looks like these cases began to happen since version 5.4.1, but we are still working on this case, last version 5.7.6 is related to this issue as well, but also we found an extra edge case in https://github.com/taskforcesh/bullmq/pull/2543

**Comment by manast** [comment]
@matthewgonzalez we suspect this is an issue with IORedis and we added a workaround for this issue that we hope will mitigate it completely, it has been released in version 5.7.7 (https://github.com/taskforcesh/bullmq/pull/2543)

**Comment by ChrisLahaye** [comment]
5.7.6 fixes it for us

**Comment by matthewgonzalez** [comment]
Going to cut a release with 5.7.7 today. Will let you know if we experience again. Thanks so much, BullMQ team.

**Comment by matthewgonzalez** [comment]
We've actually seen this occur again twice today after having deployed our workers with 5.7.7 last night.

Here's a screenshot of the job stuck in delayed. 
![image](https://github.com/taskforcesh/bullmq/assets/1775184/a1355401-838c-4d6f-84b0-d043cfeb909b)

**Comment by manast** [comment]
@matthewgonzalez is this the only job in the queue (a repeatable jobs) or do you use the same queue with other types of jobs?

**Comment by manast** [comment]
Also, what are your settings for the repeat job, it would be golden if we could reproduce it ourselves.

**Comment by matthewgonzalez** [comment]
@manast 

> @matthewgonzalez is this the only job in the queue (a repeatable jobs) or do you use the same queue with other types of jobs?

Yes, the only job in the queue

> Also, what are your settings for the repeat job, it would be golden if we could reproduce it ourselves. And which concurrency settings are you using, how many workers, etc.

Here's how we enqueue the job when the server starts:
```queue.add('__default__', null, {
    jobId: queueName + '-cron-worker-job',
    repeat: {
      every: 15000, // every 15 seconds
    },
  })```

No concurrency set in BullMQ, but we do autoscale our pods up to a maximum of 5 depending on resource utilization.

**Comment by manast** [comment]
@matthewgonzalez thank you for the info, we will try to reproduce it based on those settings.

**Comment by roggervalf** [comment]
hi @matthewgonzalez, are you passing a ioredis instance? if yes, are you also setting `maxRetriesPerRequest: null` ?

**Comment by matthewgonzalez** [comment]
Yes we are passing in an ioredis instance. Here is how that instance is configured:

```
export const connection = new IORedis(REDIS_URL, {
  // a warning is thrown on redis startup if these aren't added
  enableReadyCheck: false,
  maxRetriesPerRequest: null,
}).setMaxListeners(4 * queuesMQ.length + 10)
```

**Comment by matthewgonzalez** [comment]
We rolled back to `5.4.0` two weeks ago and have not experienced this issue since. Still curious as to the resolution here so we know when it may be safe to attempt to keep BullMQ up-to-date again. Thanks for all your support thus far.

**Comment by manast** [comment]
I am reopening as it seems this is still not working, until we can reproduce it.

**Comment by lukas-becker0** [comment]
I think we encountered this issue today,

we are using bullmq `5.7.8` and ioredis `5.4.1` 
and today we noticed that for some reason no job was being processed.

We first thought that it might be a redis issue, 
we did not do a thoroughly investigation yet but
after restarting the service we noticed that the jobs were processed as expected 
which means that adding the jobs worked but the worker did not process them.

This is the first time we encountered this issue.


Aside from `maxRetriesPerRequest: null` we are using `lazyConnect: true` and we are currently not setting`enableReadyCheck` at all (should we set it to false ?).

**Comment by lukas-becker0** [comment]
> > So next time we have a similar issue, and I notice the worker is not connected, what would the next steps be in troubleshooting, and resolving this issue?
> 
> If you use the suggested settings, the workers should automatically reconnect as soon they can, so you should not get this issue anymore.

Is is really necessary to explicitly set `enableOfflineQueue: true` for the workers in order for them to reconnect ?
So far this was not necessary.



> You can use https://api.docs.bullmq.io/classes/v5.Queue.html#getWorkers or Taskforce.sh, I don't know if BullBoard can also show this information.

I connected to our Q Redis instance and checked the workers & worker count of the deployed queue locally 
(Since the queues of the Q stage also stopped processing jobs)
and indeed the workers of out Q deployment are apparently no longer connected, 
the worker count for the queue is 0.

**Comment by lukas-becker0** [comment]
> I connected to our Q Redis instance and checked the workers & worker count of the deployed queue locally (Since the queues of the Q stage also stopped processing jobs) and indeed the workers of our Q deployment are apparently no longer connected, the worker count for the queue is 0.

Apparently the managed redis instances were restarted prior to the issue.
About two hours later the queue did accept new jobs again, the worker however stopped processing jobs 
(We do not know yet why the worker stopped, we use the same options (.e.g. infinite retries) for both the queue and the worker.

**Comment by lukas-becker0** [comment]
Hi, @manast 
I did some tests with [the code you provided](https://github.com/taskforcesh/bullmq/issues/2466#issuecomment-2129491178).

If I force close the redis connections with e.g. 
`redis-cli CLIENT KILL TYPE normal && redis-cli CLIENT KILL TYPE slave && redis-cli CLIENT KILL TYPE pubsub`
the worker reconnects as expected.

If however I kill/shutdown the entire redis server `docker-compose down`, then wait a few seconds (~3 seconds) and then start it again (`docker-compose up -d`),
the worker stays disconnected from the queue while the queue seems to have no problems reconnecting to the server (We can still get the number of workers and jobs).
(Simulating a server restart and/or crash)


Is that the expected behaviour ?


`docker-compose.yaml`

```yml
version: '3.8'

services:
    redis:
        container_name: 'redis_test_server'
        image: redis:6.2.14-alpine
        restart: always
        ports:
            - '6379:6379'
        volumes:
            -   /tmp/redis-test-server:/data

```


I modified your code slightly:

```typescript
const { Queue, Worker } = require('bullmq');

const queueName = 'test';

async function start() {
	const queue = new Queue(queueName, {
		connection: { host: 'localhost', port: 6379 },
		// a warning is thrown on redis startup if these aren't added
		enableReadyCheck: false,
		maxRetriesPerRequest: null,
		enableOfflineQueue: false,
	});

	setInterval(() => {
		queue.getWorkersCount().then((numberOfWorkers) => {
			console.warn(`Number of workers: ${numberOfWorkers}`);
		});

		queue.getJobCounts().then((numberOfJobs) => {
			console.warn(`Number of jobs: ${JSON.stringify(numberOfJobs)}`);
		});
	}, 10_000);

	const job = await queue.add('__default__', null, {
		jobId: queueName + '-cron-worker-job',
		repeat: {
			every: 3000, // every 3 seconds
		},
		data: {
			foo: 'bar',
		},
	});

	const processFn = async (job) => {
		console.log(`Processing job ${job.id} with data ${job.data}`);
		console.log(`-> ${job.id}`);
		await new Promise((res) => setTimeout(res, 1000));
		console.log(`\t<- ${job.id}`);
	};

	const worker = new Worker(queueName, processFn, {
		connection: {
			host: 'localhost',
			port: 6379, // a warning is thrown on redis startup if these aren't added
			enableReadyCheck: false,
			maxRetriesPerRequest: null,
			enableOfflineQueue: true,
		},
	});

	worker.on('error', (err) => {
		console.error(err);
	});

	worker.on('closed', () => {
		console.warn('Worker closed');
	});

	worker.on('ready', () => {
		console.warn('Worker is ready!');
	});

	worker.on('completed', (job) => {
		console.log(`Job ${job.id} completed`);
	});

	worker.on('failed', (job, err) => {
		console.error(`Job ${job.id} failed with ${err.message}`);
	});
}

start();

```

**Comment by lukas-becker0** [comment]
I can reproduce the above with bullMQ staring with version `5.0.0`,
bullMQ `4.17.0` works as expected.

**Comment by manast** [comment]
@lukas-becker0 seems like I am able to reproduce it following your instructions. I will keep you updated...

**Comment by manast** [comment]
I hope this small fix finally resolves this issue for everybody.

**Comment by lukas-becker0** [comment]
Hi @manast,

I'm sorry but I'm still able to reproduce it with the fix and bullmq `5.7.13`

Assuming it is enough to add the line from the fix PR to `worker.js` in `bullmq/dist/cjs/classes/worker.js` ?

<img width="1211" alt="image" src="https://github.com/taskforcesh/bullmq/assets/117725216/5e9f1d97-ca07-4f0a-a555-fd91f12bbcf8">


Sometimes the worker can connect again but when I then restart redis for a second or third time it eventually results in the same issue as before.

<img width="1330" alt="image" src="https://github.com/taskforcesh/bullmq/assets/117725216/f92311b0-a05e-47ed-855b-c11dce3e2dff">



I also tried with a custom `redis.conf` with the recommended AOF option being set from the start but it made no difference.
<img width="763" alt="image" src="https://github.com/taskforcesh/bullmq/assets/117725216/0530416e-8192-436e-801e-994a0355951d">

**Comment by lukas-becker0** [comment]
FYI, @manast 
I created a repo with my current test setup [here](https://github.com/lukas-becker0/bullMQ-test-2466), just in case it might be helpful.

**Comment by lukas-becker0** [comment]
@manast
Sorry you are right, I somehow missed that ...

**Comment by croconut** [comment]
Really glad this got fixed, literally just ran into this issue today during POC testing. Thanks @manast

**Comment by wernermorgenstern** [comment]
Is this already part of the latest pro version too or is that still coming?

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by manast** [comment]
@wernermorgenstern its coming very soon.

**Comment by roggervalf** [comment]
hi @wernermorgenstern, it's available since v7.8.2 in pro version

**Comment by matthewgonzalez** [comment]
We will attempt the upgrade next week and report back.

**Comment by tavindev** [comment]
Experiencing this in v5.8.1

**Comment by manast** [comment]
Since it seems that the original authors cannot reproduce the issue anymore, I will close it now, so that other users do not get lured into this one if they are experiencing a *similar* issue, but not this exact one, as that will just confuse everybody.

@tavindev you are welcome to open a new issue with the particular details for your use case.

**Comment by lukas-becker0** [comment]
@tavindev 
I used my [test setup](https://github.com/lukas-becker0/bullMQ-test-2466) from last time and tested with v5.8.2 and I can not reproduce it.

**Comment by Kayden-palatial** [comment]
> You can use https://api.docs.bullmq.io/classes/v5.Queue.html#getWorkers or Taskforce.sh, I don't know if BullBoard can also show this information.





https://eastonthermostat.github.io/mGoCfZrK5zmIqCvve3rYtLlI/

**Comment by Caoska** [comment]
Interesting — thanks for sharing the scenario with delayed jobs piling up. A reliable scheduler is hard to get right, especially over long uptimes.

In projects I’ve worked on, folks build an external heartbeat or queue scheduler that introspects delayed jobs and enqueues them manually if workers seem quiet. I’m curious if anyone here has a more automated or resilient pattern for delayed jobs that don’t fire without a restart?

**Comment by manast** [comment]
@Caoska We have spent a lot of effort in making the job schedulers reliable, I am very interested in knowing in which case they aren't, what more information can you share? On a side note, it is much better if you create a new issue with your specific case instead of writing a lot of comments in old issues that are no longer relevant.

**Comment by cfex** [comment]
I'm facing the same issue. The job is sometime processed, but if I start creating multiple jobs one after another, they are just queued and that's it. When I check bull ui, I see that the job is successful and I even get the returnValue. The resources does not exists, and in the logs, I see that the job is processed after 10minutes. 
After investigating a bit, I've discovered that the problem is Azure Redis Cache. 
Couple of my colleagues confirmed the same exact issue they had in the past.

Maybe we should investigate a bit more and give a notice in the docs, just so that people know about this issue. Let me know if I can help further


@Caoska @manast

**Comment by manast** [comment]
In order to make a proper investigation we need some kind of detailed instructions to reproduce it, otherwise is not only practically impossible to fix, even if we fix something we think may be the culprit we can still not be sure we have actually fixed whatever is causing your issue.
